provisioning/service/src/Authentication/ServiceConnectionString.cs

```diff
@@ -1,156 +1,235 @@
         /// </summary>
-        public string Audience
-        {
-            get { return HostName; }
-        }
+        public string SharedAccessKey { get; internal set; }
 
         /// <summary>
-        /// The Provisioning Service Access Policy Name
+        /// The Provisioning Service Shared Access Signature
         /// </summary>
-        public string SharedAccessKeyName
-        {
-            get;
-            private set;
-        }
+        public string SharedAccessSignature { get; internal set; }
 
         /// <summary>
-        /// The Provisioning Service Shared Access Key for the specified
-        /// access policy
+        /// The Provisioning Service Name
         /// </summary>
-        public string SharedAccessKey
+        public string ServiceName
         {
-            get;
-            private set;
+            get { return _serviceName; }
         }
 
         /// <summary>
-        /// The Provisioning Service Shared Access Signature
+        /// Returns the Provisioning Service Connection string
         /// </summary>
-        public string SharedAccessSignature
+        /// <returns></returns>
+        public override string ToString()
         {
-            get;
-            private set;
+            Validate();
+
+            StringBuilder stringBuilder = new StringBuilder();
+            stringBuilder.AppendKeyValuePairIfNotEmpty(HostNamePropertyName, HostName);
+            stringBuilder.AppendKeyValuePairIfNotEmpty(SharedAccessKeyNamePropertyName, SharedAccessKeyName);
+            stringBuilder.AppendKeyValuePairIfNotEmpty(SharedAccessKeyPropertyName, SharedAccessKey);
+            stringBuilder.AppendKeyValuePairIfNotEmpty(SharedAccessSignaturePropertyName, SharedAccessSignature);
+            if (stringBuilder.Length > 0)
+            {
+                stringBuilder.Remove(stringBuilder.Length - 1, 1);
+            }
+
+            return stringBuilder.ToString();
         }
 
-        /// <summary>
-        /// Returns the shared access signature for authorization
-        /// </summary>
-        /// <returns></returns>
-        public string GetSasToken()
+        private void Parse(string serviceConnectionString)
         {
-            string password;
-            if (string.IsNullOrWhiteSpace(SharedAccessSignature))
+            IDictionary<string, string> map = serviceConnectionString.ToDictionary(ValuePairDelimiter, ValuePairSeparator);
+
+            HostName = GetConnectionStringValue(map, HostNamePropertyName);
+            SharedAccessKeyName = GetConnectionStringOptionalValue(map, SharedAccessKeyNamePropertyName);
+            SharedAccessKey = GetConnectionStringOptionalValue(map, SharedAccessKeyPropertyName);
+            SharedAccessSignature = GetConnectionStringOptionalValue(map, SharedAccessSignaturePropertyName);
+
+            Validate();
+        }
+
+        private void Validate()
+        {
+            if (string.IsNullOrWhiteSpace(SharedAccessKeyName))
```

Comment:
Please move parameter validation within the public APIs.
1. In general the stack should be one level deep when verifying parameters.
2. This function is called several times in private functions. If something is empty but it's always required we should signal it immediately.
3. The type should be immutable - once initialized we shouldn't allow any setters.